### PR TITLE
Update CI/CD files

### PR DESCRIPTION
**Pull request by mary-cleaton**
## Description

Update miscellaneous CI/CD files. Notable points:

- New `.gitignore` uses GitHub Python default `.gitignore` then adds the last three sections in.
- Update to `.pre-commit-config.yaml` includes standardisation of lint/format (i.e. use Ruff only instead of a mix of Ruff and Black) and addition of a spellchecker (Crate-CI).
- Pull request template is changed to improve the checklists.
- Dependabot config is updated to add grouping and labelling of updates from Dependabot.
- Workflows are updated for clarity (PR workflow) or to try to fix a bug (deployment workflows).
  - I removed the bit about checking for a tag from the PyPI deployment workflow as I realised that, due to the manual deployment method, it actually can only deploy a tagged commit anyway.

The `pyproject.toml` file also has some minor updates - this is a) adding Ruff configuration and b) adding Crate-CI configuration including American English spellings to ignore (because they're variable names for packages, not things I can edit). Sadly, the one thing Crate-CI does not have is the ability to ignore incorrect spellings in a specific file or on a specific line.

## Type of change

- [x] Bug fix.
- [ ] New feature.
- [ ] Breaking change - **backwards incompatible change**, changes expected behaviour.
- [x] Non-user facing change - structural change, dev functionality, docs, etc.

## Checklist

I have:

- [x] Performed a self-review of my own code.
- [x] Commented my code appropriately, focusing on explaining my design decisions (explain why, not how).
- [x] Made corresponding changes to the documentation (comments, docstring, etc.).
- [ ] Added tests that prove my fix is effective or that my feature works.
- [x] Checked that new and existing unit tests pass locally with my changes.
- [x] Updated the change log.

No script code changes, only CI/CD code changes. Hence, no new tests.

# Reviewer

## Peer review

You should review all of the following:

- **Documentation**:
  - Are docstrings present?
  - Are comments only present where necessary and have they been updated where applicable?
- **Style guidelines**:
  - Do the new files conform to the project's contribution guidelines?
- **Functionality**:
  - Does the code work as expected?
  - Does the code handle expected edge cases and exceptions appropriately?
- **Complexity**:
  - Is the code overly complex?
  - Has the logic been split into appropriately sized functions?
- **Test coverage**:
  - Are there unit tests for all essential functions for a reasonable range of inputs and conditions?
  - Do added and existing tests pass locally on the reviewer's machine?

### Review comments

Suggestions should be tailored to the code that you are reviewing. Provide context.
Be critical and clear, but not mean. Ask questions and set actions.

<details><summary>These might include:</summary>

- Bugs that need fixing:
  - Does the new code work as expected?
  - Does the new code work with other code that it is likely to interact with?
  
- Alternative methods:
  - Could the new code be written more efficiently?
  
- Documentation improvements:
  - Does the documentation reflect how the code actually works?
  
- Additional tests that should be implemented:
  - Do the tests effectively assure that the code works correctly?
  - Are there additional edge cases/negative tests to be considered?
  
- Code style improvements:
  - Could the code be written more clearly?
  - Does the code meet the project style guide?

</details>

**Further reading: [code review best practices](https://best-practice-and-impact.github.io/qa-of-code-guidance/peer_review.html).**

### CODE REVIEW

File: .github/pull_request_template.md [hunk 1]
```diff
@@ -1,77 +1,92 @@
...
+    - Handle edge cases and exceptions.
+-   [ ] **Tests**: I confirm that unit tests:
+    - Are present where appropriate.
+    - Cover essential functions for a reasonable range of inputs and conditions.
+    - Pass locally (both new and existing tests).
+-   [ ] **Documentation**: I confirm that documentation files:
+    - Have been added/updated, as applicable.
+    - Make sense.
+    - Are not overly complex.
+-   [ ] **Style**: I confirm that all changes conform to the [contribution guide][contribution-guide].
```
Comment (danasebo) [inline .github/pull_request_template.md:64]: [contribution-guide] - broken link (also on two other places in this file)

File: .github/pull_request_template.md [hunk 2]
```diff
@@ -1,77 +1,92 @@
...
+    - Minimise complexity.
+    - Ensure all code is documented sufficiently.
+    - Ensure edge cases and exceptions are handled.
+-   **Test improvements**:
+    - Ensure sufficient coverage of both functions and inputs/conditions.
+    - Ensure all tests pass.
+-   **Style improvements**:
+    - Ensure the `cats-training` [contribution guide][contribution-guide] is followed.
+
+[contribution-guide]: docs/contribution_guide.md
```
Comment (danasebo) [inline .github/pull_request_template.md]: docs/contribution_guide.md -> /docs/contribution_guide.md

File: .github/pull_request_template.md [hunk 3]
```diff
@@ -1,77 +1,92 @@
...
+-   **Code improvements**:
+    - Ensure correct functioning.
+    - Minimise complexity.
+    - Ensure all code is documented sufficiently.
+    - Ensure edge cases and exceptions are handled.
+-   **Test improvements**:
+    - Ensure sufficient coverage of both functions and inputs/conditions.
+    - Ensure all tests pass.
+-   **Style improvements**:
+    - Ensure the `cats-training` [contribution guide][contribution-guide] is followed.
```
Comment (danasebo) [inline .github/pull_request_template.md]: there's no mention of `cats-training` in the contribution guide

File: docs/contribution_guide.md [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+- Provide well-documented, easy-to-understand code, including clear variable and function names as well as
+  explanatory comments where necessary.
+
+- Keep lists of package dependencies in [`pyproject.toml`][pyproject] up-to-date, splitting dependencies into those
+  required by all users versus developers.
+
+- Maintain unit test coverage of >=80%.
+
+- Use linting and formatting:
+  - Lint all code with `ruff` and format all code with `black`.
```
Comment (danasebo) [inline docs/contribution_guide.md]: `black` is obsolete now?

File: CHANGELOG.md [hunk 5]
```diff
@@ -48,10 +48,14 @@ and this project adheres to [semantic versioning](https://semver.org/spec/v2.0.0
 - Branch and Deploy Guide:
   - Updated description of GitHub Actions.
   - Updated deployment instructions.
-- Build method, from using `setuptools` with `setup.py`, `setup.cfg` and `requirements.txt` to using `hatchling` with `pyproject.toml`.
+- Build method, from using `setuptools` with `setup.py`, `setup.cfg` and `requirements.txt` to using `hatchling` with
+  `pyproject.toml`.
 - Location of config template, so it is included in build.
-- Minor phrasing changes to User Guide (`docs/user_guide.md`) and pull request template (`.github/pull_request_template.md`).
-- Changed all third-party reusable workflows in GitHub Actions workflows to pin to latest release's long commit hash.
+- User Guide (`docs/user_guide.md`) and Contribution Guide (`docs/contribution_guide.md`), to make minor phrasing improvements.
```
Comment (danasebo) [inline CHANGELOG.md]: I think it's worth stating that contribution guide file was also renamed (not only improved)

File: docs/contribution_guide.md [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+  overly clever one-liners.
+
+> **Note:** If your function raises any exceptions, make sure to include a `Raises` section in the docstring describing
+> the exception type and when it is triggered.
+
+### Function checklist
+
+| Requirement                                            | Check before submitting                                                             |
+| ------------------------------------------------------ | ----------------------------------------------------------------------------------- |
+| Function signatures contain type hints.                | Clearly specify input and output types. Do not duplication type hints in docstring. |
```
Comment (danasebo) [inline docs/contribution_guide.md]: "Do not duplication type hints in docstring." should probably be "Do not **duplicate** type hints in docstring."?

File: docs/contribution_guide.md [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+| House format is used.                 | Only `pytest` is used. The test has clear arrange, act, assert and (if applicable) tidy up steps.             |
+| Tests are clearly named.              | `test_` prefix is used; name clearly relates to the function being tested.                                    |
+| Test functions are clear and focused. | There is one scenario per set of parametrized values. The scenario is clearly described using `ids` argument. |
+| Fixtures are used where appropriate.  | Fixtures are used to reduce duplication.                                                                      |
+| Assertions are clear and meaningful.  | Assertions clearly show expected versus actual results.                                                       |
+| Edge cases are covered.               | Tests aim for full coverage, not just low-hanging fruit. Where tests are not performed, this is justified.    |
+
+[gitflow]: https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow
+[commits]: https://www.markdownguide.org/basic-syntax/#links
```
Comment (danasebo) [inline docs/contribution_guide.md]: These two links are not used in this doc - are they needed here?

File: .github/workflows/pull_request.yaml [hunk 8]
```diff
@@ -30,20 +26,20 @@ jobs:
...
-        uses: actions/checkout@3d3c42e5aac5ba805825da76410c181273ba90b1
-      
+        uses: actions/checkout@9c091bb21b7c1c1d1991bb908d89e4e9dddfe3e0
+
       - name: Set up Python
         id: pre-commit-python-set-up
-        uses: actions/setup-python@5fda3b95a4ea91299a34e894583c3862153e4b97
-      
+        uses: actions/setup-python@ece7cb06caefa5fff74198d8649806c4678c61a1
+
```
Comment (danasebo) [inline .github/workflows/pull_request.yaml:42]: is no python version needed here? Like on L65 in this file where it is specified for the test job

File: .github/workflows/deploy_to_test_pypi.yaml [hunk 9]
```diff
@@ -47,6 +48,6 @@ jobs:
 
       - name: Publish release distributions on PyPI
         id: publish-to-pypi
-        uses: pypa/gh-action-pypi-publish@dc37677b2e1c63e2034f94d8a5b11f265b73ba33
+        uses: pypa/gh-action-pypi-publish@cef221092ed1bacb1cc03d23a2d87d1d172e277b
         with:
           repository-url: https://test.pypi.org
```
Comment (danasebo) [inline .github/workflows/deploy_to_test_pypi.yaml:53]: This might be ok but double checking - https://github.com/pypa/gh-action-pypi-publish/blob/unstable/v1/README.md#trusted-publishing guidance say `repository-url: https://test.pypi.org/legacy/` not `repository-url: https://test.pypi.org` so please check

File: .github/workflows/deploy_to_pypi.yaml [hunk 10]
```diff
@@ -20,18 +20,15 @@ jobs:
     steps:
       - name: Check out repository
         id: build-repo-checkout
-        uses: actions/checkout@3d3c42e5aac5ba805825da76410c181273ba90b1    # only allow tagged commits to be deployed
-        if: ${{ !startsWith(github.ref, "refs/tags") }}
-        run: |
-          echo "ERROR: You can only deploy a tagged commit to PyPI."
-          exit 1
+        uses: actions/checkout@9c091bb21b7c1c1d1991bb908d89e4e9dddfe3e0    # only allow tagged commits to be deployed
```
Comment (danasebo) [inline .github/workflows/deploy_to_pypi.yaml]: probably `# only allow tagged commits to be deployed` comment can be removed now?

File: pyproject.toml [hunk 11]
```diff
@@ -67,10 +67,59 @@ changelog = "https://github.com/ONSdigital/scalelink/blob/develop/CHANGELOG.md"
 [tool.hatch.build.targets.wheel]
 only-include = ["scalelink"]
 
-[tool.isort]
-profile = "black"
-known_local_folder = "scalelink"
-skip = [".gitignore"]
+[tool.ruff]
+extend-include = ["*.ipynb", ".md"]
```
Comment (danasebo) [inline pyproject.toml]: ".md" -> "*.md"?

File: pyproject.toml
Comment (danasebo) [inline pyproject.toml]: should `isort` & `black` dependencies be removed?

File: CHANGELOG.md [hunk 12]
```diff
@@ -48,10 +48,14 @@ and this project adheres to [semantic versioning](https://semver.org/spec/v2.0.0
...
+- All third-party reusable workflows in GitHub Actions workflows, to pin to latest release's long commit hash.
+- Pull request template (`.github/pull_request_template.md`), to improve checklists and phrasing.
+- Miscellaneous CI/CD files (`.gitibnore`, `.pre-commit-config.yaml`, `.github/dependabot.yml` and whole contents of
```
Comment (danasebo) [inline CHANGELOG.md]: `.gitibnore` -> `.gitignore`

**Review by danasebo — CHANGES_REQUESTED** [review]
A few minor changes requested (typos, broken links, removal of not used packages, etc.). Once addressed, I'm happy to approve. Many thanks!

File: .github/pull_request_template.md (hunk 1 above)
Comment (mary-cleaton) [reply .github/pull_request_template.md:64]: Fixed.

File: .github/pull_request_template.md (hunk 2 above)
Comment (mary-cleaton) [reply .github/pull_request_template.md]: Fixed.

File: .github/pull_request_template.md (hunk 3 above)
Comment (mary-cleaton) [reply .github/pull_request_template.md]: Fixed. This was left in from copy-paste from a different repo.

File: docs/contribution_guide.md (hunk 4 above)
Comment (mary-cleaton) [reply docs/contribution_guide.md]: Well spotted. Fixed.

File: CHANGELOG.md (hunk 5 above)
Comment (mary-cleaton) [reply CHANGELOG.md]: Done, but also have done a rephrase of the unreleased section of the change log. This is for a) consistency and b) to ensure all files mentioned are mentioned by name.

File: docs/contribution_guide.md (hunk 6 above)
Comment (mary-cleaton) [reply docs/contribution_guide.md]: Fixed.

File: docs/contribution_guide.md (hunk 7 above)
Comment (mary-cleaton) [reply docs/contribution_guide.md]: Removed.

File: .github/workflows/pull_request.yaml (hunk 8 above)
Comment (mary-cleaton) [reply .github/workflows/pull_request.yaml:42]: I don't think it matters which Python version the pre-commits run on. They should work regardless. In contrast, the unit tests very specifically must pass on the versions of Python that the `scalelink` package requires (currently 3.10 and 3.11). Hence the difference in specification.

File: .github/workflows/deploy_to_test_pypi.yaml (hunk 9 above)
Comment (mary-cleaton) [reply .github/workflows/deploy_to_test_pypi.yaml:53]: Yeah, I was a bit confused by the `https://test.pypi.org/legacy` recommendation on that page, too. I've gone with `https://test.pypi.org/project/scalelink/` since that's where the actual Test PyPI repo sits: https://test.pypi.org/project/scalelink/

However, who knows if it'll work or not! I think this might just be one we'll have to try and see.

File: .github/workflows/deploy_to_pypi.yaml (hunk 10 above)
Comment (mary-cleaton) [reply .github/workflows/deploy_to_pypi.yaml]: Woops, yes! Removed.

File: pyproject.toml (hunk 11 above)
Comment (mary-cleaton) [reply pyproject.toml]: Fixed.

File: pyproject.toml
Comment (mary-cleaton) [reply pyproject.toml]: Removed.

File: CHANGELOG.md (hunk 12 above)
Comment (mary-cleaton) [reply CHANGELOG.md]: Fixed.

File: .github/pull_request_template.md (hunk 1 above)
Comment (danasebo) [reply .github/pull_request_template.md]: I would prefer you to use an absolute path `/docs/contribution_guide.md` (such as `[config-template]: /scalelink/configs_template.ini` in the `docs/user_guide.md` guide) rather than a relative path, if you agree.

File: .github/pull_request_template.md (hunk 2 above)
Comment (danasebo) [reply .github/pull_request_template.md]: it is a relative path now (`../docs/contribution_guide.md`) - can it be change to absolute path so it's safer and more consistent with other guides?

File: pyproject.toml [hunk 13]
```diff
@@ -122,4 +120,5 @@ locale = "en-gb"
 labeled = "labeled"
 sav = "sav"
 synchronize = "synchronize"
+"tring" = "tring"
```
Comment (danasebo) [inline pyproject.toml]: 1. why "tring" is included?
2. it should be `tring = "tring"` rather than `"tring" = "tring"`

**Review by danasebo — CHANGES_REQUESTED** [review]
Thanks Mary for your updates and additional changes - only a quick follow-up on my previous comment and one new comment on the latest changes from me before this PR can be approved. Many thanks!

File: .github/pull_request_template.md (hunk 1 above)
Comment (mary-cleaton) [reply .github/pull_request_template.md]: You're definitely right - absolute paths are better than relative. I had just forgotten how to do them when I replied to your comment, so thanks for providing the example. I hope I've got it right!

File: .github/pull_request_template.md (hunk 2 above)
Comment (mary-cleaton) [reply .github/pull_request_template.md]: Fixed (see above).

File: pyproject.toml (hunk 13 above)
Comment (mary-cleaton) [reply pyproject.toml]: Fixed. `tring` is included as it's a string in the Sorensen-Dice coefficient unit test (calculating the coefficient between `string` and `tring`).

File: .github/pull_request_template.md (hunk 1 above)
Comment (danasebo) [reply .github/pull_request_template.md]: It doesn't work now - it should be `/docs/contribution_guide.md` (not `/scalelink/docs/contribution_guide.md`)

File: pyproject.toml [hunk 14]
```diff
@@ -13,6 +13,22 @@ dependencies = [
   "raz_client >= 1.1.0",
   "rdsa_utils >= 0.16.0",
 ]
+[dependency-groups]
```
Comment (danasebo) [inline pyproject.toml]: This change breaks PR workflow run unit tests

File: .github/pull_request_template.md (hunk 2 above)
Comment (danasebo) [reply .github/pull_request_template.md]: it's probably not helpful to have two comments for the same issue! I'll close this one and keep only the one above as it's still not resolved.

File: .github/pull_request_template.md (hunk 1 above)
Comment (mary-cleaton) [reply .github/pull_request_template.md]: Drat. Evidently I still hadn't remembered the exact format for absolute paths! Thanks for the pointer. Fixed now.

File: pyproject.toml (hunk 14 above)
Comment (mary-cleaton) [reply pyproject.toml]: Sorry, this was me rushing. Forgot that that workflow uses command-line install of the test dependencies.

Fixed now!

**Comment by mary-cleaton** [comment]
Have fixed the above and also added a cooldown of 14 days for dependencies (via `uv`) and dependency updates (via Dependabot - note this automatically excludes security updates from the cooldown).

For more info, see:
- https://blog.yossarian.net/2025/11/21/We-should-all-be-using-dependency-cooldowns
- https://blog.yossarian.net/2025/12/13/cooldowns-redux

**Review by danasebo — APPROVED** [review]
Thanks Mary for all the updates and changes, looks good to me!
